src/lp_info.rs

```diff
@@ -0,0 +1,628 @@
+        }
+    }
+
+    // Handle messages to remember the last pinged album
+    pub async fn handle_message<C: BaseClient>(
+        &self,
+        client: &C,
+        ctx: &Context,
+        msg: &Message,
+    ) {
+        let msg_txt: &str = &msg.content;
+
+        // Check if the specified roles were mentioned
+        if msg
+            .mention_roles
+            .iter()
+            // Resolve ID to role
+            .filter_map(|rid| {
+                rid.to_role_cached(&ctx.cache).or_else(|| {
+                    eprintln!("Role {rid} not found");
+                    None
+                })
+            })
+            .any(|role| LP_ROLES.contains(&role.name.as_ref()))
+        {
+            let pl = match LPInfo::from_match_string(client, msg_txt).await {
+                Err(e) => {
+                    eprintln!("Error resolving spotify link: {}", e);
+                    return;
+                }
+                Ok(Some(pl)) => {
+                    let guild_name = match msg.guild_id {
+                        Some(guild) => guild
+                            .to_partial_guild(&ctx.http)
+                            .await
+                            .map(|guild| format!("[{}] ", &guild.name))
+                            .unwrap_or_default(),
+                        None => String::new(),
+                    };
+                    let username = &msg.author.name;
+                    let pinged = match &pl.playlist {
+                        PlaylistInfo::AlbumInfo {
+                            id, artist, name, ..
+                        } => format!("{id} ({artist} - {name})"),
+                        PlaylistInfo::PlaylistInfo { id, name, .. } => {
+                            format!("{id} ({name})")
+                        }
+                    };
+                    eprintln!("{guild_name}{username}: Pinged Listening Party: {pinged}");
+                    pl
+                }
+                Ok(None) => return,
+            };
+            // Store album/playlist in channel info
+            let mut channels = self.last_pinged.write().unwrap();
+            (*channels).insert(msg.channel_id, pl);
+        };
+    }
+
+    // Set the Listening party as started
+    pub fn start_lp(&self, channel: &ChannelId) {
+        let now = chrono::offset::Utc::now();
+        let mut channels = self.last_pinged.write().unwrap();
+        channels
+            .entry(*channel)
+            .and_modify(|lp_info| lp_info.started = Some(now));
+    }
+}
+
+#[async_trait]
+impl Module for ModLPInfo {
+    async fn add_dependencies(
+        builder: HandlerBuilder,
+    ) -> anyhow::Result<HandlerBuilder> {
+        builder.module::<Spotify>().await
+    }
+
+    fn register_event_handlers(&self, handlers: &mut events::EventHandlers) {
+        // Set listening party as started when ReadyPoll finishes countdown
+        let this = self.clone();
+        handlers.add_handler(move |ReadyPollStarted { channel }| {
```

Comment:
Not sure how useful the event system will be if the callbacks are synchronous. Maybe use async callbacks with tokio::spawn?

Reply:
* If the emitting code simply `await`s the callback then emitting an event can involve unbounded delays, that seems undesirable
* Instead, it could - as you suggest - `spawn` the callback as a new task. However, `spawn` itself is a synchronous function
* That doesn't seem to be practically different to calling `spawn` directly in the event handler as necessary
* I have in fact implemented the asynchronous callback version, and it comes with downsides:
  * More complicated interface, `Fn(&E) -> ()` becomes `Fn(&E) -> BoxFuture<'static, ()>`
  * mandatory `Box::pin(async {})` at the end of callback even if the handler is synchronous 

Compare:
async callback:
```rust
    fn register_event_handlers(&self, handlers: &mut events::EventHandlers) {
        let that = self.clone();
        // Set listening party as started when ReadyPoll finishes countdown
        handlers.add_handler(move |ReadyPollStarted { channel }| {
            let c = *channel;
            let this = that.clone();
            Box::pin (async move {
                this.start_lp(&c).await;
            })
        });
    }
```

sync callback:
```rust 
    fn register_event_handlers(&self, handlers: &mut events::EventHandlers) {
        let that = self.clone();
        // Set listening party as started when ReadyPoll finishes countdown
        handlers.add_handler(move |ReadyPollStarted { channel }| {
            let c = *channel;
            let this = that.clone();
            tokio::spawn (async move {
                this.start_lp(&c).await;
            });
        });
    }

```

Reply:
On second though, no,
* tokio is an implementation detail, should probably not leak here
* async callbacks is what we "mean"
*